### PR TITLE
Fix for multiplication optimization issue #367

### LINKED ISSUE #367: Multiplication optimization issue
Hi,
  first of all, thanks for the truly great compiler.

I think I stepped into a situation where the optimization yields incorrect results (or in any case, different from those I get without `-Oi`).

Here is a sample program:
```c
#include <stdio.h>

void main(void)
{
    unsigned int y=192;
    unsigned int d=y&0xFFF8;
    unsigned int e=d*32+d*8;
    unsigned int f=d*40;
    
    printf("value 1: %d value 2: %d", e, f);
}
```
I can compile it in two different ways:
```sh
#!/bin/bash
cc65 -Oi -t c64 bug.c
ca65 bug.s
ld65 -o bug -t c64 bug.o c64.lib

cc65 -t c64 bug.c
ca65 bug.s
ld65 -o bug_not -t c64 bug.o c64.lib
```
With `-Oi`, the two `*32` and `*8` multiplications are (correctly) obtained by means of shifts; but, it seems that something is wrong when the results are put together.

With no optimization, the result is as follows:
~~~~~
value 1: 7680 value 2: 7680
~~~~~
With `-Oi`, the results differ:
~~~~~
value 1: 3072 value 2: 7680
~~~~~

BTW, here is the revision of cc65 I am employing:
~~~~~
$ git log -1
commit 1fc2dfb64f6a02c30be199f22b1bd7fe123e3847
Merge: 5878155 1b4a7e3
Author: Oliver Schmidt <ol.sc@web.de>
Date:   Tue Dec 27 22:27:00 2016 +0100

    Merge pull request #363 from kylophone/master
    
    doc/ca65: fix typo
~~~~~

Hope it helps!
Cheers, Davide

### PR DESCRIPTION
**Pull request by IrgendwerA8**
Code change supplied by Uz himself.
I've added the matching test.
Fixed #367.

### CODE REVIEW

File: src/cc65/coptstop.c [hunk 1]
```diff
@@ -1832,14 +1832,16 @@ static int PreCondOk (StackOpData* D)
         return 0;
     }
 
-    /* Determine the zero page locations to use */
-    if ((D->UsedRegs & REG_PTR1) == REG_NONE) {
+    /* Determine the zero page locations to use. We've tracked the used
+     * ZP locations, so try to find some for us that are unused.
```
Comment (oliverschmidt) [inline src/cc65/coptstop.c]: Please adjust to the current multiline-comment style.

File: src/cc65/coptstop.c [hunk 2]
```diff
@@ -2009,21 +2011,32 @@ unsigned OptStackOps (CodeSeg* S)
...
 
                 /* Finalize the load info */
                 FinalizeLoadInfo (&Data.Lhs, S);
                 FinalizeLoadInfo (&Data.Rhs, S);
 
-                /* If the Lhs loads do load from zeropage, we have to include
-                ** them into UsedRegs registers used. The Rhs loads have already
-                ** been tracked.
+                /* Check if the lhs loads from zeropage. If this is true, these
+                 * zero page locations have to be added to ZPUsage, because
```
Comment (oliverschmidt) [inline src/cc65/coptstop.c]: Please adjust to the current multiline-comment style.

**Comment by IrgendwerA8** [comment]
>Do I get it right that Uz does cc65 these days?

Not really.

>Where?

In my inbox if I ask kindly enough and prepare some background info...  :smirk:

File: src/cc65/coptstop.c [hunk 3]
```diff
@@ -1833,8 +1833,8 @@ static int PreCondOk (StackOpData* D)
     }
 
     /* Determine the zero page locations to use. We've tracked the used
-     * ZP locations, so try to find some for us that are unused.
-     */
+    ** ZP locations, so try to find some for us that are unused.
+    **/
```
Comment (greg-king5) [inline src/cc65/coptstop.c]: Put only one `*` there:
```c
/*
**
*/
```

**Comment by IrgendwerA8** [comment]
Just to mention: **No need to bother him, this was an exception.**

**Comment by oliverschmidt** [comment]
Thanks for the Uz-explanation :-)

**Pull request merged**